Fixed imports, comments, and some bugs in printing

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Clean up of imports and documentation. Some syntax error-type of bugs were also fixed(?). At least no syntax errors anymore...

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
   * MathML printer now has a working fallback for Basic
   * Fortran printer for FunctionPrototype does not give a syntax error anymore (although an exception as multiline prining is not yet supported)
   * Fixed PyCode printer for sequences
<!-- END RELEASE NOTES -->

oscargus committed:
Fixed imports, comments, and some bugs

oscargus commented on sympy/printing/mathml.py:
```diff
@@ -868,7 +865,7 @@ def _print_Basic(self, e):
         mi.appendChild(self.dom.createTextNode(self.mathml_tag(e)))
         mrow.appendChild(mi)
         for arg in e:
-            x.appendChild(self._print(arg))
```
x was not defined. Not sure how to reach this in an example.

oscargus replied:
I think I've figured it out now. Will try to add something over the weekend.

oscargus replied:
Yes, this is the one I talked about below.

oscargus commented on sympy/printing/pretty/stringpict.py:
```diff
@@ -15,7 +15,7 @@
 
 from .pretty_symbology import hobj, vobj, xsym, xobj, pretty_use_unicode
-from sympy.core.compatibility import string_types, range
+from sympy.core.compatibility import string_types, range, unicode
```
unicode was used, but not imported

oscargus commented on sympy/printing/pycode.py:
```diff
@@ -488,11 +484,10 @@ class NumPyPrinter(PythonCodePrinter):
         "General sequence printer: converts to tuple"
         # Print tuples here instead of lists because numba supports
         #     tuples in nopython mode.
-        delimite.get('delimiter', ', ')
```
delimite not define, should probably have been something like `delimiter = kwargs.get(...`

Should be reasonable to add a test for.

oscargus replied:
I added a test for this. However, it does not seem like any existing code uses it.

oscargus commented on sympy/printing/rcode.py:
```diff
@@ -179,13 +178,13 @@ def _print_NegativeInfinity(self, expr):
         rhs = expr.rhs
         # We special case assignments that take multiple lines
         #if isinstance(expr.rhs, Piecewise):
+        #    from sympy.functions.elementary.piecewise import Piecewise
```
Moved it here so that it is just to uncomment the whole section.

oscargus commented on sympy/printing/glsl.py:
```diff
@@ -199,7 +198,7 @@ def _print_Function_with_args(self, func, func_args):
                 func = cond_func
             else:
                 for cond, func in cond_func:
-                    if cond(args):
```
`args` not defined, Not sure how to reach that code, but will look more into it.

oscargus replied:
Checking the code more carefully, I do not think it can be reached with the current code. It seems to be some sort of fallback in case the function is not already defined, or as some intermediate thing for future extensions.

First, the code checks if the function is in the list of known functions, if it is it gets the glsl name of the function. If that is not a string, it goes to the piece of code I changed. However, at the moment it is always a string.

Still, the updated code would not give an error when executed.

oscargus commented on sympy/printing/fcode.py:
```diff
@@ -712,7 +712,7 @@ def _print_FunctionPrototype(self, fp):
             "{function_head}\n"
             "end function\n"
             "end interface"
-        ).format(function_head=self._head(entity, fp, *args))
```
`args` not defined. Should be possible to at least cover the code in a test (although the test basically will be "does not set a variable is undefined")

oscargus replied:
I have confirmed that the old version gives:

...
  File "C:\Users\Oscar\sympy\sympy\printing\fcode.py", line 715, in _print_FunctionPrototype
    ).format(function_head=self._head(entity, fp, *args))

NameError: name 'args' is not defined
...

The new version leads to:

...
  File "C:\Users\Oscar\sympy\sympy\printing\fcode.py", line 479, in _print_Declaration
    raise NotImplementedError("F77 init./parameter statem. req. multiple lines.")

NotImplementedError: F77 init./parameter statem. req. multiple lines.
...

So I guess not much point in writing a test before that is implemented?

jksuom replied:
> writing a test before that is implemented?

There are also tests for raising NotImplementedError (in polys/tests), possibly in expectation of future implementation.

oscargus replied:
Yes, I thought about that as well. Seems to not be the actual purpose of writing tests expecting an exception, but I can for sure add one with relevant comments. Probably should add an issue for multiline printing as well (unless there already is one, will search).

oscargus commented:
I commented on the source of bugs. For some it is reasonable to add tests and I will do that.

oscargus committed:
Merge branch 'master' into cleanupprinting

oscargus committed:
Added tests

oscargus committed:
Fixed more bugs and added tests

oscargus commented:
I've added tests and fixed a few more related issues. However, I am not sure what the correct output should be. 

Example, `Basic(1, 2)`, either:

`<mrow><mi>basic</mi><mn>1</mn><mn>2</mn></mrow>`

or

`<mrow><mi>basic</mi><mn>1</mn><mn>2</mn></mrow>`

where the first one is the one currently committed (which seems to make sense from what I read).

oscargus committed:
Removed copy-paste test

oscarbenjamin commented:
This all looks good to me. I would merge but you mentioned that you would add something else and I'm not sure if you have yet...

oscargus commented:
I have added the mentioned test. However, I am not sure if the mathml output above is correct. Still, it is better than an error...

oscarbenjamin reviewed (commented):
> I've added tests and fixed a few more related issues. However, I am not sure what the correct output should be.

I've tested these using the mathml/latex boxes here: [http://www.wiris.com/editor/demo/en/developers](http://www.wiris.com/editor/demo/en/developers)

> Example, `Basic(1, 2)`, either:
> 
> `<mrow><mi>basic</mi><mn>1</mn><mn>2</mn></mrow>`

This seems to be missing the comma.

> or
> 
> `<mrow><mi>basic<mn>1</mn><mn>2</mn></mi></mrow>`

This gives an error.

oscargus commented:
Thanks for pointing out the link. Didn't think that far to actually try to find somewhere to test it...

I tried for atan2 (which is one of the few functions I know that has two arguments) and sympy gives:
`<mrow><mi>arctan</mi><mfenced><mfrac><mn>12</mn><mn>2</mn></mfrac></mfenced></mrow>`

which is sort of similar. The problem is that it converts it to a standard atan...

Probably, in this case, one would like to add brackets and the , and get something like 
`<mrow><mi>basic</mi><mfenced><mn>1</mn><mo>,</mo><mn>2</mn></mfenced></mrow>`

Ahh! Putting `<mfenced>` around it leads to , inbetween so the "correct" way is probably:

`<mrow><mi>basic</mi><mfenced><mn>1</mn><mn>2</mn></mfenced></mrow>`

That should be manageable to fix!

oscargus committed:
Corrected Basic printing

oscargus commented:
Given that the tests pass it should be good to go now. The MathML output is as correct as it can be as far as I can tell.

oscarbenjamin commented:
Should the mathml output be Basic rather than basic?

oscargus commented:
Good question! @Subhash-Saurabh who wrote this initially thought that 
https://github.com/sympy/sympy/blob/798401affb6dc657e65239334debd155bb473873/sympy/printing/mathml.py#L504
was a good line when converting a class name into a descriptor/tag.

I guess it makes sense since class names always starts with capital letters and mathematical functions only rarely does that. (Main purpose of print_Basic is, I assume, to allow any function derived from Basic to print.)

oscarbenjamin commented:
What are the examples where it makes sense to convert to lower case?

oscargus commented:
I was thinking of functions such as sin etc, where the class probably should be called Sin from a programming perspective. However, I can admit that I cannot figure out a function that is not implemented where this holds. Rather I tend to think of special functions named by their inventor...

Not a big deal for me to change the line if that is what makes sense. One may even think of adding a property of Basic that is the "print name" of the subclass. Currently, the name is determined as `n = e.__class__.__name__` followed by converting to lower case then.

One can also consider that this method has probably not been used for quite some time, if ever...

oscarbenjamin commented:
It's probably best not to change anything if we don't know any example where it should change.

I think this is fine to merge as is.
